### ISSUE TITLE
Generic param mismatch doesn't cause type error

### ISSUE THREAD

**Issue by leshow** [Microsoft/TypeScript#24170]
**TypeScript Version:**  2.8.1

I have "strict": true on and "strictFunctionTypes": true. So the bivariant behaviour (if that even applies here) should be disabled.

**Search Terms:** 
'phantom type', 'generic mismatch'
**Code**

```ts
interface Weight<T> {
  weight: number;
}

class Lb {}

function inLbs(): Weight<Lb> {
  return {
    weight: 2
  };
}

class Kg {}

function inKgs(): Weight<Kg> {
  return {
    weight: 5
  };
}

function add<T>(a: Weight<T>, b: Weight<T>): Weight<T> {
  return { weight: a.weight + b.weight };
}

function shouldNotCompile(): number {
  const w = add(inLbs(), inKgs());
  return w.weight;
}
```

**Expected behavior:**
const w = add... should throw a type error. The type of T should be either Lb or Kg. Flow correctly identifies the error: 
"^ Cannot call `add` with `inKgs()` bound to `b` because `Kg`  is incompatible with `Lb` in type argument `T`"

**Actual behavior:**
Compiles fine.
**Playground Link:**  
[link](https://www.typescriptlang.org/play/#src=%0D%0Ainterface%20Weight%3CT%3E%20%7B%0D%0A%20%20weight%3A%20number%3B%0D%0A%7D%0D%0A%0D%0Aclass%20Lb%20%7B%7D%0D%0A%0D%0Afunction%20inLbs()%3A%20Weight%3CLb%3E%20%7B%0D%0A%20%20return%20%7B%0D%0A%20%20%20%20weight%3A%202%0D%0A%20%20%7D%3B%0D%0A%7D%0D%0A%0D%0Aclass%20Kg%20%7B%7D%0D%0A%0D%0Afunction%20inKgs()%3A%20Weight%3CKg%3E%20%7B%0D%0A%20%20return%20%7B%0D%0A%20%20%20%20weight%3A%205%0D%0A%20%20%7D%3B%0D%0A%7D%0D%0A%0D%0Afunction%20add%3CT%3E(a%3A%20Weight%3CT%3E%2C%20b%3A%20Weight%3CT%3E)%3A%20Weight%3CT%3E%20%7B%0D%0A%20%20return%20%7B%20weight%3A%20a.weight%20%2B%20b.weight%20%7D%3B%0D%0A%7D%0D%0A%0D%0Afunction%20shouldNotCompile()%3A%20number%20%7B%0D%0A%20%20const%20w%20%3D%20add(inLbs()%2C%20inKgs())%3B%0D%0A%20%20return%20w.weight%3B%0D%0A%7D%0D%0A)
**Related Issues:**
